src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedSearch.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Storage;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.NavigateTo
+{
+    using CachedIndexMap = ConcurrentDictionary<(IChecksummedPersistentStorageService service, DocumentKey documentKey, StringTable stringTable), AsyncLazy<SyntaxTreeIndex?>>;
+
+    internal abstract partial class AbstractNavigateToSearchService
+    {
+        /// <summary>
+        /// Cached map from document key to the (potentially stale) syntax tree index for it we use prior to the 
+        /// full solution becoming available.  Once the full solution is available, this will be cleared to drop
+        /// all cached data.
+        /// </summary>
+        private static CachedIndexMap? s_cachedIndexMap = new();
+
+        /// <summary>
+        /// String table we use to dedupe common values while deserializing <see cref="SyntaxTreeIndex"/>s.  Like
+        /// <see cref="s_cachedIndexMap"/> it is dropped once the full solution is available.
+        /// </summary>
+        private static StringTable? s_stringTable = new();
```

Comment:
Does this still need to be `StringTable?` ?

Reply:
Oh I see, this is written to null in another file. Worth a comment on when this can be null and why

Reply:
i moved the nulling out to here.  i'll be explicit about null meaning 'dropped'.